Add switches for blue current integration.

NickKoepr opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added three switches to the blue current integration for changing the plug and charge, linked charge cards only and block settings on a charge point. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39404
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

NickKoepr committed:
Block and plug and charge switches are working correctly.

Still need to make the Link charge cards only button update when the value changes from another env.

NickKoepr committed:
Working on writing tests for the switches.

Committing in between.

NickKoepr committed:
Updated tests and added a new test.

The integration has now 100% code coverage.

NickKoepr committed:
Removed code that was commented out.

NoRi2909 commented on homeassistant/components/blue_current/strings.json:
```diff
@@ -124,6 +124,17 @@
+    },
+    "switch": {
+      "plug_and_charge": {
+        "name": "Plug and charge"
```
See my comment above to change this to ["Plug & Charge" (from ISO 15118) ](https://en.wikipedia.org/wiki/ISO_15118#Plug_&_Charge)

NickKoepr commented on homeassistant/components/blue_current/const.py:
```diff
@@ -8,3 +8,14 @@
+PLUG_AND_CHARGE = "plug_and_charge"
+VALUE = "value"
+PERMISSION = "permission"
+CH_STATUS = "CH_STATUS"
+CH_SETTINGS = "CH_SETTINGS"
```
CH_STATUS and CH_SETTINGS are object names returned by the websocket, to differentiate the different websocket messages.

joostlek replied:
Okay, but can we give them a better name than `CH_*`? As in, in HA we have config constants (`CONF_*`), attributes (`ATTR_*`), so CH_ looks weird and device/service specific

NickKoepr replied:
Ah okay, I have changed it to `CHARGEPOINT_*`.

NickKoepr committed:
Applied received feedback.

* Switch logic is moved to smaller functions that can be linked to BlueCurrentSwitchEntityDescription's.
* Changed naming for some labels.
* Moved icons to icons.json
* Removed some unnecessary code.

NickKoepr committed:
Tests now use the registered entities.

Test where failing because of the translation changes. Now, all entities are retrieved.

joostlek commented on homeassistant/components/blue_current/switch.py:
```diff
@@ -0,0 +1,182 @@
+            self.evse_id, self.connector
+        )
+        self._attr_is_on = new_state[0]
+        self.has_value = new_state[1]
```
why do we keep track on if we have a value?

NickKoepr replied:
In some scenarios a user cannot use the switches, for example when a charge point is offline. When this occurs, the  switches are disabled (so has_value is set to False then).

joostlek commented on homeassistant/components/blue_current/switch.py:
```diff
@@ -0,0 +1,182 @@
+        update_lastest_data=lambda evse_id, connector: update_on_value_and_activity(
+            PLUG_AND_CHARGE, evse_id, connector
+        ),
+        on_switch_update=lambda evse_id,
+        connector,
+        new_switch_value: update_charge_point(
+            PLUG_AND_CHARGE, evse_id, connector, new_switch_value
+        ),
```
please wrap multiline lamdas in parentehsis

joostlek commented on homeassistant/components/blue_current/switch.py:
```diff
@@ -0,0 +1,182 @@
+    update_lastest_data: Callable[[str, Connector], tuple[bool, bool]]
+    """Update the switch based on the latest data received from the websocket. The first returned boolean is _attr_is_on, the second one has_value."""
+
+    on_switch_update: Callable[[str, Connector, bool], None] | None = None
```
`turn_on_off_fn` sounds a bit better here

joostlek commented on homeassistant/components/blue_current/switch.py:
So why do we have one function to actually do the call to blue current, and one to update the data here?

NickKoepr committed:
Applied received feedback.

NickKoepr committed:
Removed now unused on_switch_update function.

NickKoepr committed:
Applied feedback on the tests for blue_current.

* Removed if statements from the test.
* Updates are not happening via the client mock instead of direct using the internal functions.
* Imported entity_registry as test fixture.
